pgmpy/inference/Sampling.py

```diff
@@ -132,17 +142,30 @@ def rejection_sample(self, evidence=None, size=1):
         """
         if evidence is None:
             return self.forward_sample(size)
-        sampled = DataFrame(columns=self.topological_order)
+        types = [(var_name, 'int') for var_name in self.topological_order]
+        sampled = np.zeros(0, dtype=types).view(np.recarray)
         prob = 1
-        while len(sampled) < size:
-            _size = int(((size - len(sampled)) / prob) * 1.5)
+        i = 0
+        while i < size:
+            _size = int(((size - i) / prob) * 1.5)
             _sampled = self.forward_sample(_size)
+
             for evid in evidence:
-                _sampled = _sampled[_sampled.ix[:, evid[0]] == evid[1]]
+                _sampled = _sampled[_sampled[evid[0]] == evid[1]]
+
             prob = max(len(_sampled) / _size, 0.01)
-            sampled = sampled.append(_sampled)
-        sampled.reset_index(inplace=True, drop=True)
-        return sampled[:size]
+
+            if HAS_PANDAS is True:
```

Comment:
@khalibartan If we just use strings and let's say we keep the default
string to pandas and the user doesn't have pandas installed. In that case
what should we do? Should we throw an error or fall back to recarray? And
if we are doing the falling back thing its the same as having None. And I
don't think throwing an error is a good way of handling this. Your opinion ?

On Aug 26, 2016 10:15 AM, "Utkarsh" <notifications@github.com> wrote:

> In pgmpy/inference/Sampling.py
> <https://github.com/pgmpy/pgmpy/pull/719#discussion_r76380706>:
>
> >              prob = max(len(_sampled) / _size, 0.01)
> > -            sampled = sampled.append(_sampled)
> > -        sampled.reset_index(inplace=True, drop=True)
> > -        return sampled[:size]
> > +
> > +            if HAS_PANDAS is True:
>
> @ankurankan <https://github.com/ankurankan> @yashu-seth
> <https://github.com/yashu-seth> I did the same earlier in continuous
> sampling :P . I think we should use string only. Not setting default value
> to None instead to recarray. This gives a sure behaviour. Setting the
> parameter to 'None' and then returning two different types based on
> installation is kind of ambiguous. In case of only string as a parameter
> user is sure what he is doing but in the later case (one with None) he must
> be careful if he forgets to set parameter. And I cannot see any advantages
> of setting it to None.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pgmpy/pgmpy/pull/719/files/badea046953ee593d9eed00ddc36a2fc288d9a9b#r76380706>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZMzY2_X7HpNxgNGjFO7YOLoFais_HPks5qjqC4gaJpZM4JnfIu>
> .
>
> --
> You received this message because you are subscribed to the Google Groups
> "pgmpy" group.
> To unsubscribe from this group and stop receiving emails from it, send an
> email to pgmpy+unsubscribe@googlegroups.com.
> To post to this group, send email to pgmpy@googlegroups.com.
> To view this discussion on the web visit https://groups.google.com/d/
> msgid/pgmpy/pgmpy/pgmpy/pull/719/r76380706%40github.com
> <https://groups.google.com/d/msgid/pgmpy/pgmpy/pgmpy/pull/719/r76380706%40github.com?utm_medium=email&utm_source=footer>
> .
> For more options, visit https://groups.google.com/d/optout.
>